### PR TITLE
Rollout retry

### PR DESCRIPTION
**Pull request by strailov**

### CODE REVIEW

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java [hunk 1]
```diff
@@ -730,4 +752,38 @@ public void triggerNextGroup(final long rolloutId) {
         startNextRolloutGroupAction.exec(rollout, latestRunning);
     }
 
+    private TargetCount calculateTargetsAndFilterBasedOnRetriedRollout(final String targetFilter, final Long createdAt, final Long dsTypeId) {
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java:755]: improper name = could be not based on retried

Comment (avgustinmm) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java:755]: Don't see need to have different method if it adds new type TargetCount (again not correctly reflects the class content)

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java [hunk 2]
```diff
@@ -730,4 +752,38 @@ public void triggerNextGroup(final long rolloutId) {
...
+        } else {
+            totalTargets = targetManagement.countByFailedInRolloutAndCompatible(
+                RolloutHelper.getIdFromRetriedTargetFilter(targetFilter), dsTypeId, createdAt);
+            baseFilter = targetFilter;
+        }
+
+        return new TargetCount(totalTargets, baseFilter);
+    }
+
+    private static class TargetCount {
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java:770]: could be record

File: hawkbit-rest/hawkbit-mgmt-resource/src/main/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtRolloutMapper.java [hunk 3]
```diff
@@ -132,6 +132,18 @@ static RolloutCreate fromRequest(final EntityFactory entityFactory, final MgmtRo
                 .weight(restRequest.getWeight());
     }
 
+    static RolloutCreate fromRetriedRollout(final EntityFactory entityFactory, final Rollout rollout) {
+        return entityFactory.rollout().create()
+                .name(rollout.getName().concat("_retried"))
```
Comment (avgustinmm) [inline hawkbit-rest/hawkbit-mgmt-resource/src/main/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtRolloutMapper.java:137]: better "_retry"

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 4]
```diff
@@ -796,6 +819,19 @@ public long countByRsqlAndCompatible(final String targetFilterQuery, final Long
...
     }
 
+    @Override
+    public long countByFailedInRolloutAndCompatible(final String rolloutId, final Long dsTypeId, final Long createdAt) {
+        final List<Specification<JpaTarget>> specList = Arrays.asList(
+            TargetSpecifications.failedActionsForRollout(rolloutId),
+            TargetSpecifications.isCompatibleWithDistributionSetType(dsTypeId));
+
+        if (createdAt != null) {
+            specList.add(TargetSpecifications.createdBefore(createdAt));
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java:829]: why createdBefore is used - rollout fail filter would not contain devices created after anyway

File: hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/repository/RolloutHelper.java [hunk 5]
```diff
@@ -253,4 +256,12 @@ public static void checkIfRolloutCanStarted(final Rollout rollout, final Rollout
                     + rollout.getStatus().name().toLowerCase());
         }
     }
+
+    public static boolean isRolloutRetried(final String targetFilter) {
+        return targetFilter.contains("failedRollout");
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/repository/RolloutHelper.java:261]: is this case sensitive? 
aren't lower cased attributes are supported at the moment? e.g. "createdat"
if search attributes are case insensitive this check shall be also such.
If not - better use lowercase as in other cases

**Review by avgustinmm — CHANGES_REQUESTED** [review]

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java (hunk 1 above)
Comment (strailov) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java:755]: Different method was needed in order to escape duplication ...

File: hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/repository/RolloutHelper.java [hunk 6]
```diff
@@ -253,4 +256,12 @@ public static void checkIfRolloutCanStarted(final Rollout rollout, final Rollout
...
+    }
+
+    public static String getIdFromRetriedTargetFilter(final String targetFilter) {
```
[the remark this answers is not in the document]
Comment (strailov) [reply hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/repository/RolloutHelper.java:264]: In normal Target filter - yes, spaces are supported. But in this case this target filter is special and is set by the system - normal user cannot set this target filter.

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 7]
```diff
@@ -796,6 +819,15 @@ public long countByRsqlAndCompatible(final String targetFilterQuery, final Long
         return JpaManagementHelper.countBySpec(targetRepository, specList);
     }
 
+    @Override
+    public long countByFailedInRolloutAndCompatible(final String rolloutId, final Long dsTypeId) {
+        final List<Specification<JpaTarget>> specList = Arrays.asList(
+            TargetSpecifications.failedActionsForRollout(rolloutId),
+            TargetSpecifications.isCompatibleWithDistributionSetType(dsTypeId));
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java:826]: aren't the failed targets already checked for compatibility?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 8]
```diff
@@ -715,6 +727,17 @@ public long countByRsqlAndNotInRolloutGroupsAndCompatible(final Collection<Long>
         return JpaManagementHelper.countBySpec(targetRepository, specList);
     }
 
+    @Override
+    public long countByFailedRolloutAndNotInRolloutGroupsAndCompatible(Collection<Long> groups, String rolloutId,
+        DistributionSetType distributionSetType) {
+        final List<Specification<JpaTarget>> specList = Arrays.asList(
+            TargetSpecifications.failedActionsForRollout(rolloutId),
+            TargetSpecifications.isNotInRolloutGroups(groups),
+            TargetSpecifications.isCompatibleWithDistributionSetType(distributionSetType.getId()));
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java:736]: aren't the failed targets already checked for compatibility?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 9]
```diff
@@ -693,6 +693,18 @@ public Slice<Target> findByTargetFilterQueryAndNotInRolloutGroupsAndCompatible(f
         return JpaManagementHelper.findAllWithoutCountBySpec(targetRepository, pageRequest, specList);
     }
 
+    @Override
+    public Slice<Target> findByFailedRolloutAndNotInRolloutGroupsAndCompatible(Pageable pageRequest, Collection<Long> groups,
+        String rolloutId, DistributionSetType distributionSetType) {
+        final List<Specification<JpaTarget>> specList = Arrays.asList(
+            TargetSpecifications.failedActionsForRollout(rolloutId),
+            TargetSpecifications.isNotInRolloutGroups(groups),
+            TargetSpecifications.isCompatibleWithDistributionSetType(distributionSetType.getId())
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java:702]: aren't the failed targets already checked for compatibility?

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java [hunk 10]
```diff
@@ -278,6 +293,26 @@ Slice<Target> findByTargetFilterQueryAndNotInRolloutGroupsAndCompatible(@NotNull
...
+     * and that are not assigned to one of the retried {@link RolloutGroup}s and are
+     * compatible with the passed {@link DistributionSetType}.
+     *
+     * @param pageRequest
+     *            the pageRequest to enhance the query for paging and sorting
+     * @param groups
+     *            the list of {@link RolloutGroup}s
+     * @param rolloutId
+     *            rolloutId of the rollout to be retried.
+     * @param distributionSetType
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java:307]: not a parameter anymore

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java [hunk 11]
```diff
@@ -296,6 +331,23 @@ Slice<Target> findByTargetFilterQueryAndNotInRolloutGroupsAndCompatible(@NotNull
...
+    /**
+     * Counts all targets with failed actions for specific Rollout
+     * and that are not assigned to one of the {@link RolloutGroup}s and are
+     * compatible with the passed {@link DistributionSetType}.
+     *
+     * @param groups
+     *            the list of {@link RolloutGroup}s
+     * @param rolloutId
+     *            rolloutId of the rollout to be retried.
+     * @param distributionSetType
```
Comment (avgustinmm) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java:343]: not parameter anymore

**Review by avgustinmm — CHANGES_REQUESTED** [review]

**Review by avgustinmm — APPROVED** [review]

**Comment by avgustinmm** [comment]
lgtm

**Pull request merged**